architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+
+**User Experience:**
+
+```bash
+# Check current Radius version
+> rad version
+RELEASE   VERSION   BICEP     COMMIT
+0.44.0    v0.44.0   0.33.93   1dd17270ec9bc9e764f314fa62c248406034edda
+
+# Perform a basic upgrade to a specific version
+> rad upgrade kubernetes --version v0.45.0
+
+Initiating Radius upgrade from v0.44.0 to v0.45.0...
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Compatible upgrade path
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✓ Applications Resource Provider
+  ✓ Controller
+  ✓ ...
+Performing post-upgrade verification...
+  ✓ All components healthy
+
+Upgrade complete! Radius has been successfully upgraded to v0.45.0.
+Note: Your local Radius CLI is still v0.44.0. To upgrade your CLI, download the v0.45.0 version.
+```
+
+**Result:**
+
+1. Pre-flight checks validate the upgrade is possible
+1. System automatically creates a user data backup for recovery
+1. All control plane components are upgraded in sequence
+1. Post-upgrade verification confirms system health
+1. User is notified about the CLI version mismatch
+
+**Exceptions:**
+
+1. If version check fails because the target is lower than current version
+2. If components fail health checks after upgrade
+
+#### Scenario 2: DevOps engineer upgrades with custom configuration
+
+**User Story:** As a DevOps engineer, I need to upgrade Radius with custom configuration parameters to match organization's infrastructure requirements.
+
+**User Experience:**
+
+```bash
+# Upgrade with custom configuration values
+> rad upgrade kubernetes --version v0.44.0 --set global.monitoring.enabled=true --set global.resources.limits.memory=2Gi
+
+Initiating Radius upgrade from v0.43.0 to v0.44.0 with custom configuration...
+Custom configuration detected:
+  - global.monitoring.enabled: true
+  - global.resources.limits.memory: 2Gi
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Compatible upgrade path
+  ✓ Custom configuration validated
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components with custom configuration:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✓ Applications Resource Provider
+  ✓ Controller
+  ✓ ...
+Applying custom configuration settings...
+Performing post-upgrade verification...
+  ✓ All components healthy
+  ✓ Custom configuration applied successfully
+
+Upgrade complete! Radius has been successfully upgraded to v0.44.0 with your custom configuration.
+```
+
+**Result:**
+
+1. System acknowledges custom configuration parameters
+1. Pre-flight checks validate both upgrade path and configuration
+1. Control plane components are upgraded with custom settings
+1. Verification confirms both health and configuration application
+
+**Exceptions:**
+
+1. If custom configuration parameters are invalid (or should it be ignored?)
```

Comment:
What's the benefit of ignoring invalid config?

Reply:
I had added that as a note to myself. After playing around with the code, we shouldn't ignore and just error out.